ARCs/arc-0060.md

```diff
@@ -0,0 +1,449 @@
+
+#### Display message
+
+The wallet **MUST** display a message for any signing request. The message **MUST** be computed dynamically taking as inputs the `metadata.scope`, the `signingData.signer`, and the message being signed `msg`.
+
+> The `message` structure is left to the implementation. An example is:
+>
+> `You are about to sign <msg> with scope <metadata.scope> and signer <signingData.signer>`.
+
+The wallet **SHOULD** display a warning in the message if the `ScopeType` is a `LSIG`.
+
+## Rationale
+
+This API was designed to enable the secure signing of structured data with Algorand wallets. The API:
+
+- Is generic to _pure_ ed25519 key pairs and does not require key manipulation
+- Is easy to extend with custom JSON schema to support future signing use cases (e.g. authentication)
+- Is secure by design preventing malicious applications from tricking users into signing malicious data or in the worst cases approving malicious transactions.
+
+This API was not designed to sign Algorand transactions or a group of transactions (see ARC-1 instead).
+
+## Backwards Compatibility
+
+This ARC is not backward compatible with [ARC-47](./arc-0047.md) Logic Signature Template.
+
+ARC-47 introduces the method `algo_templatedLsig` that takes a `LogicSignatureDescription` object and returns the signature of a logic signature program for delegation.
+
+An alternative using the ARC-60 `signData(signingData, metadata)` is given below, where:
+
+- `signingData.data` **MUST** specify:
+  - the `LogicSignatureDescription` object as detailed in ARC-47
+  - the `values` of the templated variables and the `hash` of the expected program
+  - the `ARC60Domain` must be set to the domain separator `"Program"`, as specified in the <a href="https://github.com/algorandfoundation/specs/blob/master/dev/crypto.md#domain-separation">Algorand specs</a>.
+
+- `metadata` **MUST** specify:
+  - an extended ARC-60 compatible JSON schema
+  - `ScopeType` is set to `LSIG`.
+
+The extended schema can be found in the `../assets/arc-0060` folder with the file [lsig-schema.json](../assets/arc-0060/lsig-schema.json).
+
+### Semantics for signing a Logic Signature
+
+- if the `metadata.scope` is set to `LSIG`:
+  - the `ARC60Domain` value **MUST** be equal to the byte array corresponding to the prefix `"Program"`.
+  - the JSON schema **MUST** define the `signingData.data.bytes` field as an array of objects, such as the `LogicSignatureDescription`, `values`, and `hash`.
+  - the wallet **MUST** follow the ARC-47 specification: compile the program from the received template and check its integrity with the `hash`.
+  - the compiled program **MUST** be used with the prefix to compute the signing message `msg`, as `msg="Program" || <compiled_lsig>`.
+
+## Test Cases
```

Comment:
Once the ARC is stable and about to be finalized, it'd be nice to add some test cases. We could specify a keypair, a message (according to the AUTH schema) and show the resulting signature as a test case.

In another test case we could have a keypair derived from a hd-wallet.